src/System.Net.Http/src/System/Net/Http/Headers/KnownHeader.cs

```diff
@@ -12,34 +13,36 @@ internal sealed class KnownHeader
         private readonly HttpHeaderType _headerType;
         private readonly HttpHeaderParser _parser;
         private readonly string[] _knownValues;
+        private readonly byte[] _asciiBytesWithColonSpace;
 
-        public KnownHeader(string name, HttpHeaderType headerType, HttpHeaderParser parser, string[] knownValues = null)
+        public KnownHeader(string name) : this(name, HttpHeaderType.Custom, null)
         {
             Debug.Assert(!string.IsNullOrEmpty(name));
             Debug.Assert(HttpRuleParser.GetTokenLength(name, 0) == name.Length);
-            Debug.Assert(headerType != HttpHeaderType.Custom);
-            Debug.Assert(parser != null);
-
-            _name = name;
-            _headerType = headerType;
-            _parser = parser;
-            _knownValues = knownValues;
         }
 
-        public KnownHeader(string name)
+        public KnownHeader(string name, HttpHeaderType headerType, HttpHeaderParser parser, string[] knownValues = null)
         {
             Debug.Assert(!string.IsNullOrEmpty(name));
             Debug.Assert(HttpRuleParser.GetTokenLength(name, 0) == name.Length);
 
             _name = name;
-            _headerType = HttpHeaderType.Custom;
-            _parser = null;
+            _headerType = headerType;
+            _parser = parser;
+            _knownValues = knownValues;
+
+            _asciiBytesWithColonSpace = new byte[name.Length + 2]; // + 2 for ':' and ' '
```

Comment:
In `KnownHeaders.TryGetKnownHeader`, we end up doing a comparison between the Unicode string name and the Ascii header bytes from the wire, using `ByteArrayHelpers.EqualsOrdinalAsciiIgnoreCase`.  

Since we now have the header name in Ascii here, it might be more efficient to do an Ascii to Ascii comparison there.

Reply:
Yes, for that overload.

Reply:
We don't currently have an EqualsOrdinalAsciiIgnoreCase equivalent for two span arguments (though we should in the near future).  I could add one, but I don't think it'll make a significant difference perf-wise, as the helper is already just doing an ordinal comparison, indexing into the string and the span.  We can revisit subsequently as desired.

Reply:
That's a good point; I forgot this was IgnoreCase.

I do wonder (for future investigation) whether it would be beneficial to do a case-sensitive compare first, since that will usually succeed.  Especially if Span is optimized to do a memcmp-like compare (do you know if it is?).

Reply:
> do you know if it is?

It is:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/SpanHelpers.byte.cs#L889